File: impl/app/android/app/src/main/java/com/ak/contexta/domain/usecase/ResendPendingAlertsUseCase.kt
package com.ak.contexta.domain.usecase

import android.util.Log
import com.ak.contexta.domain.AppInfoProvider
import com.ak.contexta.domain.DeveloperAlertSender
import com.ak.contexta.domain.ErrorContext
import com.ak.contexta.domain.error.AppError
import com.ak.contexta.domain.error.LlmFatalCode
import com.ak.contexta.domain.error.StructuralCode
import com.ak.contexta.domain.repository.ArticleRepository
import com.ak.contexta.domain.time.TimeProvider
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import javax.inject.Inject
import javax.inject.Singleton

/**
 * 补发未送达的飞书告警。
 *
 * 实时告警是尽力而为：生成期间进程可能被系统终止（后台冻结/Job 超时），
 * `sendArticleFailure` / `sendBatchReady` 的 HTTP 调用可能未完成就随进程消失。
 * 错误事件已落库（generation_error_log），批次状态已落库（article_batch），
 * 但通知标记（notified_at / ready_notified_at）仍为 null —— 启动时补发这些告警。
 *
 * 幂等保证：
 * - 告警发送成功才回写 notified_at（SQL 带 `AND notified_at IS NULL` 条件，只写一次）
 * - 发送失败（返回 false）不回写，下次启动继续补发
 */
@Singleton
class ResendPendingAlertsUseCase @Inject constructor(
    private val articleRepository: ArticleRepository,
    private val alertSender: DeveloperAlertSender,
    private val timeProvider: TimeProvider,
    private val appInfo: AppInfoProvider
) {
    companion object {
        private const val TAG = "ResendPendingAlerts"
        /** 只补发最近 24 小时内的错误，更早的视为已处理完毕（人工可见于错误历史）。 */
        private const val LOOKBACK_MS = 24 * 60 * 60 * 1000L
    }

    suspend operator fun invoke() {
        val now = timeProvider.nowMillis()
        val sinceIso = Instant.ofEpochMilli(now - LOOKBACK_MS)
            .atZone(ZoneId.systemDefault())
            .format(DateTimeFormatter.ISO_OFFSET_DATE_TIME)

        var resent = 0
        // 1. 补发未通知的错误告警
        val unnotifiedErrors = articleRepository.getUnnotifiedErrors(sinceIso)
        for (error in unnotifiedErrors) {
            val context = ErrorContext(
                batchId = if (error.entityType == "BATCH") error.entityId
                else articleRepository.getArticle(error.entityId)?.batchId,
                articleId = if (error.entityType == "ARTICLE") error.entityId else null,
                appVersion = appInfo.versionCode,
                timestamp = now
            )
            val sent = when (error.errorCode) {
                "LLM_FATAL" -> alertSender.sendLlmFatalError(
                    AppError.LlmFatal(LlmFatalCode.AUTH_FAILED, error.errorMessage, null),
                    context
                )
                "STRUCTURAL_PIPELINE_BLOCKED", "PIPELINE_BLOCKING" -> alertSender.sendStructuralError(
                    AppError.Structural(StructuralCode.UNEXPECTED_ERROR, error.errorMessage, null),
                    context
                )
                else -> alertSender.sendArticleFailure(
                    status = if (error.errorCode == "LLM_RECOVERABLE_EXHAUSTED") "FAILED" else "TIMEOUT",
                    errorCode = error.errorCode,
                    errorMessage = error.errorMessage,
                    context = context
                )
            }
            if (sent) {
                articleRepository.markErrorNotified(error.id)
                resent++
            }
        }

        // 2. 补发未通知的批次完成告警
        val unnotifiedBatches = articleRepository.getReadyBatchesUnnotified()
        for (batch in unnotifiedBatches) {
            val articleCount = articleRepository.getArticles(batch.id).size
            val sent = alertSender.sendBatchReady(
                batchId = batch.id,
                articleCount = articleCount,
                batchGeneratedOn = batch.generatedOn,
                batchDifficulty = batch.difficultyLevelSnapshot,
                context = ErrorContext(batch.id, null, appInfo.versionCode, now)
            )
            if (sent) {
                articleRepository.markBatchReadyNotified(batch.id)
                resent++
            }
        }

        if (resent > 0) {
            Log.i(TAG, "Resent $resent pending alert(s): ${unnotifiedErrors.size} error(s), ${unnotifiedBatches.size} batch(es)")
        }
    }
}

File: impl/app/flutter/lib/domain/usecase/resend_pending_alerts_usecase.dart
import '../../core/time/iso8601.dart';
import '../app_info_provider.dart';
import '../developer_alert_sender.dart';
import '../error/app_error.dart';
import '../repository/article_repository.dart';
import '../time/time_provider.dart';

/// 补发未送达的飞书告警（对照 Kotlin ResendPendingAlertsUseCase.kt）。
///
/// 实时告警是尽力而为：生成期间进程可能被系统终止（后台冻结/Job 超时），
/// sendArticleFailure / sendBatchReady 的 HTTP 调用可能未完成就随进程消失。
/// 错误事件已落库（generation_error_log），批次状态已落库（article_batch），
/// 但通知标记（notified_at / ready_notified_at）仍为 null —— 启动时补发这些告警。
///
/// 幂等保证：
/// - 告警发送成功才回写 notified_at（SQL 带 `AND notified_at IS NULL` 条件，只写一次）
/// - 发送失败（返回 false）不回写，下次启动继续补发
class ResendPendingAlertsUseCase {
  ResendPendingAlertsUseCase({
    required this._articleRepository,
    required this._alertSender,
    required this._timeProvider,
    required this._appInfo,
  });

  final ArticleRepository _articleRepository;
  final DeveloperAlertSender _alertSender;
  final TimeProvider _timeProvider;
  final AppInfoProvider _appInfo;

  /// 只补发最近 24 小时内的错误，更早的视为已处理完毕（人工可见于错误历史）。
  static const int _lookbackMs = 24 * 60 * 60 * 1000;

  Future<void> call() async {
    final now = _timeProvider.nowMillis();
    final sinceIso =
        isoOffsetDateTime(DateTime.fromMillisecondsSinceEpoch(now - _lookbackMs));

    // 1. 补发未通知的错误告警
    final unnotifiedErrors = await _articleRepository.getUnnotifiedErrors(sinceIso);
    for (final error in unnotifiedErrors) {
      final batchId = error.entityType == 'BATCH'
          ? error.entityId
          : (await _articleRepository.getArticle(error.entityId))?.batchId;
      final context = ErrorContext(
        batchId: batchId,
        articleId: error.entityType == 'ARTICLE' ? error.entityId : null,
        appVersion: _appInfo.versionCode,
        timestamp: now,
      );
      final sent = switch (error.errorCode) {
        'LLM_FATAL' => await _alertSender.sendLlmFatalError(
            LlmFatal(
              code: LlmFatalCode.authFailed,
              message: error.errorMessage,
            ),
            context,
          ),
        'STRUCTURAL_PIPELINE_BLOCKED' || 'PIPELINE_BLOCKING' =>
          await _alertSender.sendStructuralError(
            Structural(
              code: StructuralCode.unexpectedError,
              message: error.errorMessage,
            ),
            context,
          ),
        _ => await _alertSender.sendArticleFailure(
            status:
                error.errorCode == 'LLM_RECOVERABLE_EXHAUSTED' ? 'FAILED' : 'TIMEOUT',
            errorCode: error.errorCode,
            errorMessage: error.errorMessage,
            context: context,
          ),
      };
      if (sent) {
        await _articleRepository.markErrorNotified(error.id);
      }
    }

    // 2. 补发未通知的批次完成告警
    final unnotifiedBatches = await _articleRepository.getReadyBatchesUnnotified();
    for (final batch in unnotifiedBatches) {
      final articleCount = (await _articleRepository.getArticles(batch.id)).length;
      final sent = await _alertSender.sendBatchReady(
        batchId: batch.id,
        articleCount: articleCount,
        batchGeneratedOn: batch.generatedOn,
        batchDifficulty: batch.difficultyLevelSnapshot,
        context: ErrorContext(
          batchId: batch.id,
          articleId: null,
          appVersion: _appInfo.versionCode,
          timestamp: now,
        ),
      );
      if (sent) {
        await _articleRepository.markBatchReadyNotified(batch.id);
      }
    }
  }
}
